src/java.desktop/share/classes/com/sun/java/swing/SwingUtilities3.java

```diff
@@ -135,4 +143,73 @@ public static RepaintManager getDelegateRepaintManager(Component
         }
         return delegate;
     }
+
+    @FunctionalInterface
+    public interface PaintInterface {
+        void paintUnscaledBorder(Component c, Graphics g, int x, int y,
+                                 int w, int h, double scale, int strokeWidth);
+    }
+
+    public static void paintBorder(Component c, Graphics g, int x, int y,
+                            int w, int h, PaintInterface paintFunction) {
+
+        //STEP 1: RESET TRANSFORM
+        //save old values
+        AffineTransform at = null;
+        Stroke oldStk = null;
+        boolean resetTransform = false;
+        int stkWidth = 1;
+        double scaleFactor = 1;
+
+        if (g instanceof Graphics2D g2d) {
+            at = g2d.getTransform();
+            scaleFactor = at.getScaleX();
+            oldStk = g2d.getStroke();
+
+            // if m01 or m10 is non-zero, then there is a rotation or shear
+            // skip resetting the transform
+            resetTransform = ((at.getShearX() == 0) && (at.getShearY() == 0));
+
+            if (resetTransform) {
+                g2d.setTransform(new AffineTransform());
```

Comment:
Can we keep the original comment about this:
                /* Deactivate the HiDPI scaling transform,
                 * so we can do paint operations in the device
                 * pixel coordinate system instead of the logical coordinate system.
                 */

Reply:
Sure. I can add it.

Reply:
I believe @rajamah meant to put this comment before resetting the transform to explain why it's done.

https://github.com/openjdk/jdk/blob/9911405e543dbe07767808bad88534abbcc03c5a/src/java.desktop/share/classes/javax/swing/border/LineBorder.java#L165-L169

The original comment above for `resetTransform` variable could also be preserved, it explains why we don't reset the transform when shear and rotation are used.